ethcore/src/miner/stratum.rs

```diff
@@ -0,0 +1,198 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Client-side stratum job dispatcher and mining notifier handler
+
+use ethcore_stratum::{JobDispatcher, RemoteWorkHandler, PushWorkHandler};
+use std::sync::{Arc, Weak};
+use std::sync::atomic::Ordering;
+use std::thread;
+use nanoipc;
+use util::{H256, U256, FixedHash, H64, clean_0x};
+use ethereum::ethash::Ethash;
+use ethash::SeedHashCompute;
+use util::{Mutex, RwLock};
+use miner::{Miner, MinerService};
+use client::Client;
+use block::IsBlock;
+use std::str::FromStr;
+use rlp::encode;
+
+/// IPC socket dedicated to stratum
+pub const STRATUM_SOCKET_NAME: &'static str = "parity-stratum.ipc";
+/// IPC socket for job dispatcher
+pub const JOB_DISPATCHER_SOCKET_NAME: &'static str = "parity-mining-jobs.ipc";
+
+/// Job dispatcher for stratum service
+pub struct StratumJobDispatcher {
+	last_work: RwLock<Option<(H256, U256, u64)>>,
+	seed_compute: Mutex<SeedHashCompute>,
+	client: Weak<Client>,
+	miner: Weak<Miner>,
+}
+
+impl JobDispatcher for StratumJobDispatcher {
+	fn initial(&self) -> Option<String> {
+		// initial payload may contain additional data, not in this case
+		self.job()
+	}
+
+	fn job(&self) -> Option<String> {
+		self.with_core(|client, miner| {
+			if let Some((pow_hash, difficulty, number)) = miner.map_sealing_work(&*client, |b| {
+				let pow_hash = b.hash();
+				let number = b.block().header().number();
+				let difficulty = b.block().header().difficulty();
+
+				(pow_hash, *difficulty, number)
+			}) {
+				*self.last_work.write() = Some((pow_hash, difficulty, number));
+				Some(self.payload(pow_hash, difficulty, number))
+			} else { None }
+		})
+	}
+
+	fn submit(&self, payload: Vec<String>) {
+		if payload.len() != 3 {
+			warn!(target: "stratum", "submit_work: invalid work submit request({:?})", payload);
+		}
+		else {
+			trace!(target: "stratum", "submit_work: {} {} {}", &payload[0], &payload[1], &payload[2]);
+		}
+
+		let nonce = match H64::from_str(clean_0x(&payload[0])) {
+			Ok(nonce) => nonce,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work ({}): invalid nonce ({:?})", &payload[0], e);
+				return;
+			}
+		};
+
+		let pow_hash = match H256::from_str(clean_0x(&payload[1])) {
+			Ok(pow_hash) => pow_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work ({}): invalid hash ({:?})", &payload[1], e);
+				return;
+			}
+		};
+
+		let mix_hash = match H256::from_str(clean_0x(&payload[2])) {
+			Ok(mix_hash) => mix_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work ({}): invalid mix-hash ({:?})",  &payload[2], e);
+				return;
+			}
+		};
+
+		trace!(target: "stratum", "submit_work: Decoded: nonce={}, pow_hash={}, mix_hash={}", nonce, pow_hash, mix_hash);
+		let client = self.client.upgrade().unwrap();
+		let miner = self.miner.upgrade().unwrap();
```

Comment:
why have these as weak pointers if the program will panic (read: unrecoverable and unexpected error) if they aren't kept alive?